src/Components/Web.JS/src/Services/NavigationManager.ts

```diff
@@ -248,7 +248,9 @@ async function onBrowserInitiatedPopState(state: PopStateEvent) {
     await navigateHistoryWithoutPopStateCallback(delta);
   }
 
-  await notifyLocationChanged(false);
+  // We don't know if popstate was triggered for a navigation that can be handled by the client-side router,
+  // so we treat it as a intercepted link to be safe.
+  await notifyLocationChanged(/* interceptedLink */ true);
```

Comment:
So effectively what this does is:
* If this matches a route entry for the client side router, nothing changes
* If it does not match a route, fully reload the page (non-enhanced reload)

Is that correct? If so, I wonder if we could update [this line](https://github.com/dotnet/aspnetcore/blob/4498ca72286fb1464c10df0a0e203309f6cca681/src/Components/Components/src/Routing/Router.cs#L257) to call `NavigationManager.Refresh()` with the `forceReload` option set to `false`. I believe that should make it so an enhanced page update gets performed, when possible.

Reply:
Does this all mean we'd no longer ever do enhanced nav on "back"? That is, on popstate, the only two possible reactions would be "interactive nav" or "full-page reload"?

Reply:
I believe this code path only gets taken when navigation interception is enabled, so we'd still do enhanced navigation on "back" in purely statically routed scenarios.

That said, if we change the behavior of navigating "back" from an interactively routed page to a statically routed page to do an "enhanced refresh" instead of a full page reload, there could now be a problem where this code path (among other interactive routing logic) now _does_ get taken, because there wasn't a full page reload to "disable" navigation interception (maybe this is what your comment was implying already).

Addressing https://github.com/dotnet/aspnetcore/issues/50069 would likely fix that particular issue, but it may require new public API (prototyped previously in https://github.com/dotnet/aspnetcore/pull/50564). But that's probably too risky to introduce in a servicing release, and probably not necessary to add in .NET 9 given that you have another fix for this problem.

So, it seems to me like those nuances rule out the suggestion in my previous comment, meaning the question is whether the original approach is sufficient: Always perform a full page reload when transitioning from an interactively routed page to a statically routed page.

Reply:
> Always perform a full page reload when transitioning from an interactively routed page to a statically routed page.

Is this scenario is well-defined enough to state the most common examples? In general we don't really have common cases where people transition between different router rendermodes - normally the router itself is either just static or interactive forever.

It even theoretically possible for this change to surface a behavioral difference in any situation other than the one that led to the 404 that we're trying to fix?

It seems like you've thought through the end-to-end flow more completely than I have here so am hoping this is straightforwards to answer!

Reply:
> I know about the auth case where the router rendermode is determined dynamically, but I thought part of why this works is that the auth system itself forces a full-page load when it wants to issue an auth challenge,

I'm not sure I follow. Auth challenges do cause the cookie handler to trigger 302 redirect, but this bug has nothing to do with that. It's entirely caused by the way the template with individual auth changes the render mode from `InteractiveWebAssembly` or `InteractiveAuto` to `null`. The repro gif in the original issue shows the bug without the user ever attempting to visit a page that requires authorization. They merely visit the login page, but don't log in.

![repro video](https://github.com/dotnet/aspnetcore/assets/38009549/9c60c648-b194-45fb-8194-aef5e1dec64e)

Reply:
> It even theoretically possible for this change to surface a behavioral difference in any situation other than the one that led to the 404 that we're trying to fix?

Don't think so. The only place the `interceptedLink` argument gets used anywhere is in:

https://github.com/dotnet/aspnetcore/blob/4498ca72286fb1464c10df0a0e203309f6cca681/src/Components/Components/src/Routing/Router.cs#L245-L258

So, the only place this change would have any effect is when we're just about to display the "not found" page, and the change just takes the reload code path.

Reply:
Awesome. Thanks very much for tracing this all through.

Thanks to @halter73's gif above I realised the change I'm doing in #51046 actually does **not** fix this problem and instead would make it a bigger problem because now it would also happen whenever people use the new "static page" override feature. So, your fix here looks like it will address the problem not only for the existing auth-enabled templates, but also for future pages with `@attribute [StaticPage]`.

In a way this simplifies things for us because we don't need two different fix strategies (one for main, another for backport). We can just use your fix in main and backport as long as we're confident in the result.